src/RamIndexAnalysis.h

```diff
@@ -0,0 +1,344 @@
+        return orders;
+    }
+
+    /** @brief Get all chains */
+    const ChainOrderMap getAllChains() const {
+        return chainToOrder;
+    }
+
+    /** @brief check whether number of bits in k is not equal
+        to number of columns in lexicographical order */
+    bool isSubset(SearchSignature cols) const {
+        int idx = map(cols);
+        return card(cols) < orders[idx].size();
+    }
+
+    /** @brief map the keys in the key set to lexicographical order */
+    void solve();
+
+    /** @brief convert from a representation of A verticies to B verticies */
+    static SearchSignature toB(SearchSignature a) {
+        SearchSignature msb = 1;
+        msb <<= (4 * 8 - 1);
+        return (a | msb);
+    }
+
+    /** @brief convert from a representation of B verticies to A verticies */
+    static SearchSignature toA(SearchSignature b) {
+        SearchSignature msb = 1;
+        msb <<= (4 * 8 - 1);
+        return (b xor msb);
+    }
+
+protected:
+    SearchSet searches;          // set of search patterns on table
+    OrderCollection orders;      // collection of lexicographical orders
+    ChainOrderMap chainToOrder;  // maps order index to set of searches covered by chain
+    MaxMatching matching;        // matching problem for finding minimal number of orders
+
+    /** count the number of bits in key */
+    static size_t card(SearchSignature cols) {
+        size_t sz = 0, idx = 1;
+        for (size_t i = 0; i < sizeof(SearchSignature) * 8; i++) {
+            if (idx & cols) {
+                sz++;
+            }
+            idx *= 2;
+        }
+        return sz;
+    }
+
+    /** @brief maps search columns to an lexicographical order (labeled by a number) */
+    int map(SearchSignature cols) const {
+        assert(orders.size() == chainToOrder.size() && "Order and Chain Sizes do not match!!");
+        int i = 0;
+        for (auto it = chainToOrder.begin(); it != chainToOrder.end(); ++it, ++i) {
+            if (it->find(cols) != it->end()) {
+                assert((size_t)i < orders.size());
+                return i;
+            }
+        }
+        abort();
+    }
+
+    /** @brief determine if key a is a strict subset of key b*/
+    static bool isStrictSubset(SearchSignature a, SearchSignature b) {
+        auto tt = static_cast<SearchSignature>(std::numeric_limits<SearchSignature>::max());
+        return (~(a) | (b)) == tt && a != b;
+    }
+
+    /** @brief insert an index based on the delta*/
+    void insertIndex(std::vector<int>& ids, SearchSignature delta) {
+        int pos = 0;
+        SearchSignature mask = 0;
+
+        while (mask < delta) {
+            mask = SearchSignature(1 << (pos));
+            SearchSignature result = (delta) & (mask);
+            if (result) {
+                ids.push_back(pos);
+            }
+            pos++;
+        }
+    }
+
+    /** @brief given an unmapped node from set A we follow it from set B until it cannot be matched from B
```

Comment:
I think it would be worth documenting the parameters in the doxygen for this method, and even parameters in this file.